[v6.2] Server Access Section

inertial-frame opened the pull request:
Part of https://github.com/gravitational/teleport/issues/6391:

1. Added a server access icon for the new section here: https://github.com/gravitational/next/pull/223
2. Modified links in the intro page to point to the new section.
3. Covered topics described on the intro page (auditing, SSO, SSH, short-lived certs).
4. Updated TOC links and modified existing links to the articles that moved into the new section
5. Created diagrams to support new articles: https://github.com/gravitational/teleport/pull/6790
6. Verified everything in linting and set up steps (see videos) for new and SSH articles.
7. Placeholder reference file but no link for section (since it wasn't included in https://github.com/gravitational/teleport/issues/6391 above). 

Please make any and all recommendations, suggestions, areas for improvement. Thanks!

inertial-frame committed:
docs: server access

inertial-frame committed:
docs: update link

inertial-frame committed:
docs: videos

inertial-frame committed:
docs: add video to intro

inertial-frame commented:
@benarent - per the note above is there a better place for the videos to live? Also, format?

inertial-frame committed:
docs: tweaks

inertial-frame committed:
docs: typo

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,318 @@
+  
+   Most cloud providers require SSH to initially access, configure, and provision your cloud instances. We'll configure and launch our instances then demonstrate how to use the `tsh` tool and Teleport in SSH mode thereafter.
+
+3. Install certbot and acquire TLS (SSL) certificates from Let's Encrypt:
```
I will do another run, but on the first glance - you don't need a certbot when you are using teleport's acme (as below) with acme mode, Teleport itself acts as a certbot. You can remove all certbot-related entries from getting started guide.

inertial-frame replied:
Thanks! Right, I just saw that ACME is not only supported but is now optional (Teleport can handle TLS acme certs on its own) as of these nifty changes certbot: https://github.com/gravitational/teleport/pull/5178. I will update the docs here and create an issue to update the other docs.

inertial-frame committed:
docs: remove certbot

inertial-frame committed:
docs: add titles to admonitions

inertial-frame committed:
docs: edit pass

inertial-frame committed:
docs: improved

inertial-frame committed:
docs: improved intro, videos

inertial-frame committed:
docs: improved access requests

inertial-frame committed:
docs: api example expanded

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+
+   ii. Then, we'll associate TLS (SSL) certificates from Let's Encrypt with them below.
+
+   iii. Make sure ports: `9090`, `443`, `3023-3025`, and `3080` are all open.
```
For `tele.example.com` we only need ports `3080` - for web access, and `3023` for SSH proxy.

Let's explain what these ports are used for as well. https://goteleport.com/docs/admin-guide/#ports

No other ports will need to be open. In general, let's try to guide users to using as little open ports as possible, especially on the internet.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,384 @@
+3. Install Teleport on each instance.
+
+   ```bash
+   # Install Teleport
````
Let's borrow the table from https://goteleport.com/docs/installation/#linux. You can use include directive to avoid re-typing all instructions.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,384 @@
+   sudo apt-get install teleport
+   ```
+
+4. Configure Teleport to use ACME and to automatically acquire a TLS certificate:
````
Most readers are familiar with let's encrypt, so I would leed with this.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,384 @@
+     -o file
+   ```
+
+5. Launch Teleport on `auth.example.com`: `sudo teleport start`. 
````
I would use a full-line format here as opposed to inline command to keep a flow.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+
+   i. Create a user to access the `auth.example.com` Web UI: 
+
+   - `sudo tctl users add auth-admin --roles=editor,access,admin --logins=root,ubuntu,ec2-user`  
```
You only need `editor` and `access` roles, no need for `admin` role. We want readers to use the principle of the least privilege here.

klizhentas replied:
This part is not related to adding node, so might be better to factor it out of this part and use-later, or have a separate section.

inertial-frame replied:
OK, I think `admin` is required to see the actual cluster, correct? At least when I created users with lesser privileges the added nodes didn't appear. I'll verify again. That being said, the other two should be dropped if `admin` is added.

klizhentas replied:
@inertial-frame `access` should be enough to see the nodes, as long as the role has some SSH login assigned to it. In general we don't want to promote admin, because it's permissions are too overarching.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+   - `sudo tctl tokens add --type=node`  
+   - "Node" mode specifies that the Teleport Node that will join will act as an SSH server.  
+
+   iv. You should see:
```
For impatient readers I would show UI flow and the CLI flow in a tab, similar to here:

https://goteleport.com/docs/access-controls/guides/dual-authz/#access-requests-flow

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+   # Join cluster
+   sudo teleport start \
+     --roles=node \
+     --token=d62629ea83e181644e0778af22ec3a8e \
```
Let's not hardcode real tokens in our docs. Some folks will copy-paste the docs and the tokens with them. Instead you can ask readers to generate tokens and put them to files, and then the commands can read tokens from files. Take a look at how Roman is doing it in the database access guide:

https://goteleport.com/docs/database-access/guides/postgres-aws/#start-database-service-with-cli-flags

klizhentas commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,384 @@
+   iv. You should see:
+
+   ```bash
+   # Join cluster
````
You can point the joined node to `tele.example.com:3080` it will be easier to setup as well. This will spin the node in IOT mode. The node will establish the tunnel to the proxy and then proxy will use the tunnel.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+
+   iii. Make sure ports: `9090`, `443`, `3023-3025`, and `3080` are all open.
+
+2. Create two new instances of your desired Linux distribution (such as CentOS, Debian, Ubuntu).
```
We should explain what those instances will be used for: `tele.example.com` is our bastion server hosting proxy and auth, and the other server is just a node we are connecting to the proxy.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+
+   ii. You should now be able to view both Nodes at: `https://auth.example.com/web/cluster/auth.example.com/nodes` in the Web Interface after logging in with your the user `auth-admin`.
+
+   ![Both Nodes in](../../img/server-access/teleport_ui.png)
```
Let's try using new plugin for images Alexey has created:

https://github.com/gravitational/next/pull/206

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+
+   ![Both Nodes in](../../img/server-access/teleport_ui.png)
+
+## Step 2a. Use tsh to connect to the cluster
```
I would lead with the use-case that users can relate to, and then explain the tool below:

* Step 2/n SSH into the server

klizhentas commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,384 @@
+   ip-172-31-17-124 e7719967-b6eb-4fc0-8e53-00fac6d896d4 127.0.0.1:3022 env=example,hostname=ip-172-31-17-124 
+   ```
+
+4. Using the command `tsh ssh root@ip-172-31-17-124` locally, `auth-admin` can SSH into the `auth` Node and do the following:
````
I would add a tab switcher here that will use `tsh` by default but also show a bit more lengthy `ssh` command equivalent:

`ssh -J tele.example.com user@host`

This will answer a very common first question of whether users can use ssh to connect to nodes.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+Your browser does not support the video tag.
+</video>
+
+## Step 2b. Use tsh to introspect cluster
```
Searching nodes using resource catalog?

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+   - They have visibility into all resources within the cluster due to their defined and assigned role.
+   - They can also quickly view any Node or grouping of Nodes that've been assigned a `label`.
+
+2. Executing `tctl nodes ls` displays:
```
I would use `tsh ls` here - it's an equivalent command that is available in `tsh`

klizhentas commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,384 @@
+
+   ```yaml
+   authentication:
+     type: "saml"
````
`type: github`?

inertial-frame replied:
Hmm... I think the current type is borrowed from the Okta SAML YAML config. I see there's a GitHub-specific format briefly mentioned in the Admin page.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+
+   ![Users Panel](../../img/server-access/github_user_added.png)
+
+## Step 3b. Create short-lived certificates
```
I would push this out of scope of getting started guide, may be add a separate guide and use next steps to link to it. It is a common question, but we have to keep readers focused on the use-case.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+   - Team members can access private resources securely using certificates and SSH tunnels.
+   - No sharing of resource secrets, passwords, or credentials only certificates that are unique to that user and specified length of time.
+
+## Step 3c. Connect to the Teleport API
```
Same as above.

klizhentas commented on docs/pages/server-access/guides.mdx:
```diff
@@ -0,0 +1,10 @@
+
+# Guides
+
+- [SSH](./guides/teleport-ssh.mdx)
```
Try to use tile image group here:

https://teleport-docs-next-git-feature-partners-block-gravitational.vercel.app/storybook/?path=/story/docs-tile-tileimage--tile-image-group

With a new layout:

```
---
title: Server Access Guides
description: How to set up and configure Teleport Server Access for SSH and PAM
layout: tocless-doc
---

klizhentas commented on docs/pages/server-access/guides/openssh-teleport.mdx:
```diff
@@ -0,0 +1,338 @@
+---
```
I assumed that's one-to-one port so not reviewing this part.

inertial-frame replied:
No major changes - I did validate this and the other PAM doc. Some SEO and editing changes coming in.

klizhentas commented on docs/pages/server-access/guides.mdx:
```diff
@@ -0,0 +1,10 @@
+
+- [SSH](./guides/teleport-ssh.mdx)
+- [Using Teleport with PAM](./guides/ssh-pam.mdx)
+- [OpenSSH Guide](./guides/openssh-teleport.mdx)
```
Let's rename it to `openssh.mdx`. We don't need `teleport` in the URL names, because our root contains teleport: `https://goteleport.com`

klizhentas commented on docs/pages/server-access/guides/teleport-ssh.mdx:
```diff
@@ -0,0 +1,241 @@
+---
```
Same here, no need for `teleport-ssh`.  The purpose of this section is not clear compared to getting started. Is that a guide on access requests or something else? I think we need to narrow the scope of this particular guide.

klizhentas commented on docs/pages/server-access/guides/teleport-ssh.mdx:
```diff
@@ -0,0 +1,241 @@
+   # Join cluster
+   sudo teleport start \
+      --roles=node \
+      --token=d62629ea83e181644e0778af22ec3a8e \
```
same here: re token hardcodes

klizhentas commented on docs/pages/server-access/guides/teleport-ssh.mdx:
```diff
@@ -0,0 +1,241 @@
+   - The `tsh ssh` command allows one to do anything they would if they were to SSH into a server using a third-party tool. 
+   - All commands executed by `auth-admin` are recorded and can be replayed in the Teleport Web interface.
+
+## Configure Single Sign-On
```
This should be a separate SSO for SSH guide. Will give you more space to talk about the SSO specific features in more detail.

I suggest:  remove the `teleport-ssh` guide out of scope of this PR, split this guide into SSO and Access requests guides.

klizhentas commented on docs/pages/server-access/guides/teleport-ssh.mdx:
```diff
@@ -0,0 +1,241 @@
+Your browser does not support the video tag.
+</video>
+
+### Defining Events
```
This will make more sense in architecture overview section of server access, where you would explain how events work.

klizhentas commented on docs/pages/server-access/guides/teleport-ssh.mdx:
```diff
@@ -0,0 +1,241 @@
+
+Review the `client` [pkg.go reference documentation](https://pkg.go.dev/github.com/gravitational/teleport/api/client) for more information about working with the Teleport API programmatically.
+
+## Configure short-lived certificates for access
```
Same here, I would factor this into a separate guide like here:

https://goteleport.com/docs/kubernetes-access/guides/cicd/

klizhentas commented on docs/pages/server-access/guides/teleport-ssh.mdx:
```diff
@@ -0,0 +1,241 @@
+
+2. Admins also create short-lived certificates for resources users must connect to directly:
+
+   - Execute: `tctl auth sign --ttl=87600h --user=jenkins --out=jenkins.pem` to generate a certificate valid for a specified *Time To Live* (TTL).
```
let's avoid recommending long-lived certs, see the guide above.

klizhentas commented on docs/pages/server-access/introduction.mdx:
```diff
@@ -0,0 +1,53 @@
+
+## Guides
+
+- [SSH](./guides/teleport-ssh.mdx)
```
Let's see if we can use the same layout `tocless-doc` as linked above and use tiles as in our root intro with most common use-cases here:

https://goteleport.com/docs/

klizhentas reviewed (commented):
OK, we are making progress! Having lots of comments, but I think this will be worth 2-3 more iterations to get to a great structure with server access!

inertial-frame commented on docs/pages/server-access/introduction.mdx:
```diff
@@ -0,0 +1,53 @@
+
+Teleport Server Access is designed for the following kinds of scenarios:
+
+1. When up to a vast number of clusters must be managed using the command-line (`tsh`) or programmatically (through the Teleport API) and you want to simplify your stack, security, and configuration complexity.
```
Will revise this and verify other sections are similarly active. Ran these through our Flesch-Kincaid scoring and was near 100%. (It checks for voicing, apparently.)

inertial-frame commented on docs/config.json:
```diff
@@ -123,8 +139,8 @@
-        { "title": "Single sign-on (SSO)", "slug": "/enterprise/sso/ssh-sso/" },
-        { "title": "Access Requests", "slug": "/enterprise/workflow/" },
+        { "title": "Single Sign-On (SSO)", "slug": "/enterprise/sso/ssh-sso/" },
+        { "title": "Access Workflows", "slug": "/enterprise/workflow/" },
```
Yep that's correct. Will revert!

inertial-frame commented on docs/pages/server-access/introduction.mdx:
```diff
@@ -0,0 +1,53 @@
+
+- [SSH](./guides/teleport-ssh.mdx)
+- [Using Teleport with PAM](./guides/ssh-pam.mdx)
+- The [OpenSSH Guide](./guides/openssh-teleport.mdx) 
```
I put a placeholder in here in case we needed to add reference information. I'll happily remove it.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+
+1. Verify you have a registered domain name. 
+
+   i. We'll create two subdomains, one for each Linux instance: `auth.example.com` and `ssh.example.com`.
```
That's a definite improvement - I'll make sure to explain the `bastion` pattern as well. My intent was to mimic scenarios where one has multiple public resources (AWS and GCP endpoints, for example) but I should've explicated that further here as well. Thanks!

klizhentas replied:
It's a good security practice - having multiple layers of security. In this case not having SSH servers being available on the internet reduces the surface of attack. We don't have to explain it in the guide, probably reserving it for architecture sections, but setting a good practice in getting started will help

inertial-frame committed:
docs: requested changes

inertial-frame commented:
@klizhentas - Thanks for the very helpful comments, criticism, and suggestions! I believe I've addressed all issues. I will be creating several branches off this for the sections your recommended spinning off into their own articles. Also, some of the videos need to be tweaked which I will handle this afternoon.

Main items:

1. Improved layout, tiles.
2. I implemented the Bastion pattern much better and confirmed the ports, steps, etc. to set that up.
3. Major rewording and step corrections.
4. Added the recommended MDX components.

klizhentas commented:
@inertial-frame thanks! I will review tomorrow and get back to you.

klizhentas reviewed (changes requested):
I've done another pass. I focused on `getting-started` guide and introduction.

The structure is coming together.  Critical parts to address:

**Documentation structure**

Let's keep guides focused on achieving the goal and remove the features and benefits from this page. Let's think of guides as cooking recipes. Take a look at this unified documentation guide:

https://documentation.divio.com/how-to-guides/

Introduction - explains in 1-2 minutes what is a server access + video demo.
Guides are like recipes where we help user to achieve success without explaining much.
Architecture - this is where we explain how things work.

**Run the guide from start to finish in the rendered mode**

* Try to copy-paste and run every single command and make sure it works as is. If it does not, let the reader know what parameters they should modify. Most readers will be copy pasting commands and if they don't work this will be perceived as a bug.

* Let's try to avoid hard-coded values - tokens, ca-pins, etc. Teleport's mission is to deliver the most secure setup possible. It's OK for us to be opinionated, especially in guides.


* Let's keep the amount of Teleport specific terminology to a minimum and when introducing it, do it gradually.

* Try to give a guide a final pass, from start to finish in a rendered mode, not markdown mode. I've noticed that markdown looks great, e.g. `iii.`, but when rendered, does not actually work. Our readers will be using redndered web view, so we should optimize for it.

See other comments inline below.

P.S. I would not worry about lots of comments from me on this early stages, more important to get to the highest quality with 3-4 iterations of this section, other sections will go much faster.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,384 @@
+
+# Getting Started
+
+Teleport Server Access provides full SSH capabilities out-of-the-box:
```
@inertial-frame I've noticed that github sometimes hides the `mdx` files in the diff view.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,404 @@
+
+## Prerequisites
+
+- A Linux environment (such as Ubuntu 20.05, CentOS 8.0-1905, or Debian 10) within which we'll create multiple instances.
```
Hmmm .... would you prefer one instance here with two Nodes (on one bare metal)? I figured using two instances would be a better exemplar of a Bastion pattern.

inertial-frame commented on docs/pages/server-access/guides/ssh.mdx:
```diff
@@ -0,0 +1,252 @@
+---
+title: Server Access Getting Started Guide | Teleport Docs
```
Note this was requested by UI/UX for SEO. This should be implemented as a lint check.

One advantage to having this explicitly set here is that we know the resultant `title` length. Google will truncate any title > 70 which can impact search and SEO.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,404 @@
+   | - | - | - |
+   | 3023 | Proxy | SSH port clients connect to. A proxy will forward this connection to port `#3022` on the destination Node. |
+   | 3025 | Auth | SSH port used by the Auth Service to serve its API to other Nodes in a cluster. |
+   | 3080 | Proxy | HTTPS connection to authenticate `tsh` users and web users into the cluster. The same connection is used to serve a Web UI. |
```
Ah, my fault - I did test this and was unable to connect to `3080` through my web client. Hence, the addition of port `443`. I will reverify since I probably missed something.

Also, elsewhere in the docs and blog we refer people to the Home - Getting Started doc for implementing the Bastion pattern:

1. https://goteleport.com/docs/getting-started/ <- lists  `3023`, `3024`, `3025`, and `443` as being required.
2. https://goteleport.com/blog/ssh-bastion-host/ <- directs people to use the above

Should that be updated too or would you like to keep that and direct the existing blog content to this new article?

inertial-frame replied:
I think port `443` is required to add an initial user since they must accept the invite. And, this is required to tsh into the cluster later on. Is my understanding correct?

klizhentas replied:
It's either 443 or 3080 depending on the configuration. These ports are interchangeable. Let's use 443 then to be consistent.

inertial-frame replied:
Great thanks - I'll drop the `3080` and `3025` options after a bit more tinkering for the minimal scenario here.

I see in the YAML we generate:
```
listen_addr: 0.0.0.0:3025
  cluster_name: bastion.example.com
  ....
  enabled: "yes"
  listen_addr: 0.0.0.0:3023
  web_listen_addr: :443
  public_addr: bastion.example.com:443
```

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,404 @@
+## Prerequisites
+
+- A Linux environment (such as Ubuntu 20.05, CentOS 8.0-1905, or Debian 10) within which we'll create multiple instances.
+- A registered Doman Name.
```
Regarding the first comment: Apologies - I did remove a lengthier pitch statement but left a high-level overview about features. I'll remove that as well.

inertial-frame commented on docs/pages/server-access/guides/ssh.mdx:
```diff
@@ -0,0 +1,252 @@
+
+# Teleport SSH
+
+Teleport Server Access provides full SSH capabilities out-of-the-box:
```
I really appreciate you're sharing that documentation with me. It's really great, I 100% agree with their philosophy, appreciate the brevity, reviewed their Git history, and read through a lot of the metadocs. 

Generally, I agree and think we should align our content according to your vision like so:

1. Introduction - an Overview and Landing Page with video using new tiles or some equivalent.
2. Our Getting Started Guides - Tutorials going sequentially from start to finish. We need to improve the ability to complete these with zero breaks (not going to a blog post or other article to complete) and better code snippets.
3. Guides - How-To's - focused on specific features or implementations.
4. Reference - high brevity, mostly tables describing API  or CLI methods, params, flags.
5. Architecture - Concepts - high-level architectural or conceptual content.

I've synced with Ben to see about reusing or adding an original video and have some new diagrams to add to these pages that'll help illustrate some of the sequential operations. We haven't been doing that in a lot of Guides and Getting Started docs and I think that'll help.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,404 @@
+     bordered
+     caption="Subdomain and IP mapping"
+   >
+     <Image
```
I have not gotten `Figure` to work locally or otherwise.

klizhentas commented on docs/pages/server-access/guides/openssh.mdx:
```diff
@@ -0,0 +1,354 @@
+
+## Architecture
+
+<Figure
```
@iAdramelk the render of images not working for us, can you please take a look?

inertial-frame committed:
docs: new figure syntax

inertial-frame commented on docs/pages/server-access/introduction.mdx:
```diff
@@ -0,0 +1,77 @@
+3. When Teleport users require a complete, dedicated, and secure SSH option (Teleport Node running in SSH mode) and more than a certificate authority (Teleport Auth) with proxy (Teleport Proxy).
+4. When resource and network security must be maximized: SSH certificates over secret keys, Two-Factor Authentication (2FA), Single Sign-On (SSO), and short-lived certificates.
+
+Server Access features are organized into three main offerings: 
```
NP - converted the previous into tiles per:  https://github.com/gravitational/teleport/pull/6694#discussion_r625985985 - sure will remove.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,404 @@
+  This guide involves configuring your Teleport Nodes into the *Bastion* pattern so that [only a single Node can be accessed publically](https://goteleport.com/blog/ssh-bastion-host/).
+</Admonition>
+
+<Admonition type="tip" title="Tip">
```
Moving this higher up to be part of the replacement to the offering block.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,404 @@
+
+   - Close port `22` on your second, private, Linux instance now that your Node is configured and running.
+
+## Step 3/5. SSH into the server
```
I've left the user step in 2/5 so the reader can sign in to get visual confirmation that the node has been added.

inertial-frame replied:
Happy to rearrange this but I think displaying the great UI is a good way to get that visual confirmation up to that point.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,404 @@
+
+Now, we've got our cluster up and running. Let's see how easy it is to connect to our two Nodes. We can use `tsh` to SSH into the cluster:
+
+1. On your local machine, log in through `tsh`: 
```
My fault - thought that tool came prepackaged with teleport itself locally. We do indicate Teleport as a requirement but I'm happy to add independent or alternative steps for `tsh`.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,404 @@
+
+4. Configure Teleport on the *Bastion Host*.
+
+   i. Teleport will now automatically acquire a [Let's Encrypt](https://letsencrypt.org/docs/client-options/) ACME-compliant TLS certificate. 
```
This is a great improvement. We're appropriate I've converted a few numbered sequences to bullet points which I think read better.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,404 @@
+
+   iii. Create a *join token* to add the second Node to the `tele.example.com` cluster:
+
+   - `sudo tctl tokens add --type=node`  
```
Also, all examples are for dummy instances and such that are spun up and then deleted once I complete testing. Sorry should have mentioned that previously too.

inertial-frame committed:
docs: rest of changes + diagrams

inertial-frame committed:
docs: revert svg to png

inertial-frame committed:
docs: remove ssh guide link

inertial-frame committed:
docs: revert other svg to png

inertial-frame commented:
@klizhentas - Thanks for the great suggestions. I've updated the docs to reflect your changes, factored out the other guides, added the new diagrams from #6790. (I also went ahead and resolved each of the comments I addressed above given the number of comments in this thread if that's OK.)

I've also synced with @benarent to help with a video for the Introduction page. Once you give the green light on this PR, I'll proceed to other Guides which we can add prior to and after the 6.2 release on the 21st per https://github.com/gravitational/teleport/issues/6391.

inertial-frame committed:
docs: correct port number

klizhentas commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,479 @@
+     <TabItem label="Create a token">
+      ```bash
+      # Let's save the token to a file
+      sudo tctl tokens add --type=node > token.file
````
As described, this command will not work, however this will:

```
$ tctl tokens add --type=node | grep -oP '(?<=token:\s).*' > token.file
```

I recommend testing the docs by following them command by command and making sure the copied commands work.

inertial-frame replied:
Thanks for the recommendation @klizhentas - I did in fact go through and test these per your suggestion - I think the code snippet above was copied directly from your earlier comment (far above) and tested:

![Screen Shot 2021-05-12 at 7 23 44 PM](https://user-images.githubusercontent.com/82530006/118060135-b11e1f00-b357-11eb-8a94-e53faaa16eaa.png)

Am I missing something?

klizhentas replied:
This part works, but if you try to start a node with the next command, teleport start --token=./token will not work, because the token file contains full text, but not the token.

inertial-frame replied:
Yeah for the example above I switched the name to `token` for ease. I could have used `token.file` instead (as you well know) with the same result. The CLI is indifferent to the suffix only the supplied token path.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,479 @@
+  </TabItem>
+</Tabs>
+
+## Step 4/5. Use tsh and the universal resource catalog to introspect the cluster
```
Let's remove universal.

inertial-frame replied:
Just want to verify - I should remove universal? And, not use the full product name?

inertial-frame replied:
I've corrected the typo here to `unified` since it still appears in the marketing guidelines. However, happy to alter this pending other branding changes.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,479 @@
+   sudo tctl nodes ls
+   ```
+
+   It displays the universal resource catalog with all queried resources in one view: 
````
Same here

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,404 @@
+
+   - `tsh ssh root@env=example ls` will execute the `ls` command on each Node displaying the contents of each to the screen.
+
+## Step 5/5. Connect to the Teleport API
```
Ping on removing the API step out of getting started and to the separate guide.

klizhentas commented on docs/pages/server-access/guides/openssh.mdx:
```diff
@@ -0,0 +1,348 @@
+  allowFullScreen
+/>
+
+## Architecture
```
Let's call this overview. Real architecture will be in the architecture section.

inertial-frame replied:
Also updated commands to use sudo and removed bash $. Stylistically it looks nice but practically it reduces getting through the guides.

klizhentas commented on docs/pages/server-access/guides/openssh.mdx:
```diff
@@ -0,0 +1,348 @@
+  bordered
+  caption="Teleport OpenSSH Recording Proxy"
+>
+  ![Teleport OpenSSH Recording Proxy](../../../img/openssh-proxy.svg)
```
In this diagram, it's not clear what's encrypted and what's decrypted for me. You can replace with `Terminate SSH` (record) -> Reestablish SSH

and `Forward agent` and it will help readers familiar with open ssh to understand how it works.

inertial-frame replied:
This is being handled here: https://github.com/gravitational/web/issues/1774 but I'll include an updated placeholder.

klizhentas reviewed (changes requested):
Almost there. The getting started flow is great, but you still need to remove `API` guide into a separate (and a very brief) guide. There are also a couple of glitches with commands that need fixing.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,479 @@
+   | Port | Service | Description |
+   | - | - | - |
+   | 3023 | Proxy | SSH port clients connect to. A proxy will forward this connection to port `#3022` on the destination Node. |
+   | 3024 | Proxy	SSH port used to create "reverse SSH tunnels" from behind-firewall environments into a trusted proxy server. |
```
Thanks for drawing this to my attention. I did in fact read that while going through the docs and from the comments above. However, in testing:

Without `3024` open:

![Screen Shot 2021-05-12 at 8 05 25 PM](https://user-images.githubusercontent.com/82530006/118062878-8e8f0480-b35d-11eb-9e03-c892becdd998.png)

With it open:

![Screen Shot 2021-05-12 at 8 08 14 PM](https://user-images.githubusercontent.com/82530006/118062998-d2820980-b35d-11eb-9a69-18a4be72a20a.png)

---

The only difference between those two scenarios was leaving port `3024` open. I tried multiple permutations.

inertial-frame commented on docs/pages/server-access/introduction.mdx:
```diff
@@ -0,0 +1,38 @@
+description: Teleport Server Access features and introduction.
+---
+
+# Server Access
```
Our docs use this standard inconsistently. Some use the Header, others don't - should we standardize this going forward? Does the Header get generated only for Introduction or all documents. I think this isn't itself documented in the https://goteleport.com/docs/docs/ documentation. Thanks!

inertial-frame replied:
Tested this with the h1 meta item and it did not render. I've kept this in as a result if that's OK.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,479 @@
+1. Creating a Teleport Auth Node cluster we'll add a Teleport SSH Node to.
+2. SSH into the cluster using Teleport.
+3. Introspecting the cluster using Teleport features.
+4. Create a simple Go client to interact with the Teleport API.
```
I will remove the step and the Go client into a Guide.

inertial-frame replied:
Updated this to https://github.com/gravitational/teleport/pull/6873

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,479 @@
+- A Two-Factor Authentication app (such as Authy or Google Authenticator).
+- Teleport (=teleport.version=) installed locally.
+
+## Overview
```
Ah, apologies - got the approval to use this here. I'll remove the diagrams that aren't appropriate for the article and sections.

inertial-frame commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,479 @@
+      ```
+     </TabItem>
+
+     <TabItem label="token.file">
````
OK, just to understand more clearly - you wanted a Tab flow here, correct? Thanks and sorry for any confusion.

inertial-frame committed:
docs: requested features

inertial-frame committed:
Merge branch 'branch/v6' into adam/docs/62accsect

inertial-frame committed:
docs: update images

inertial-frame committed:
docs: remove video temporarily

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,347 @@
+## Prerequisites
+
+- Two instances of your favorite Linux environment (such as Ubuntu 20.05, CentOS 8.0-1905, or Debian 10).
+- A registered Doman Name with an available subdomain like `tele.example.com`.
```
@inertial-frame `a registered domain name` (see the typo in the name)

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,347 @@
+
+3. Configure Teleport on the *Bastion Host*.
+
+   Teleport will now automatically acquire an ACME-compliant TLS certificate. 
```
Nit: the certificate itself is not `ACME-compliant`, it's just a usual X.509 web certificate teleport gets through ACME protocol. So may be change this to `Teleport will now automatically acquire an X.509 certificate using ACME protocol`

inertial-frame replied:
Thanks - that could have been improved. The IETF https://datatracker.ietf.org/doc/html/rfc8555:

```
The guiding use case for ACME is obtaining certificates for websites.
```

I had intended this to mean compliant with ACME itself per:

```
Let's Encrypt uses the ACME protocol to verify that you control a given domain name and to issue you a certificate. To get a Let's Encrypt certificate, you'll need to choose a piece of ACME client software to use. 
```

Corrected.

klizhentas commented on docs/pages/server-access/getting-started.mdx:
```diff
@@ -0,0 +1,347 @@
+
+   | Port | Service | Description |
+   | - | - | - |
+   | 3023 | Proxy | SSH port clients connect to. A proxy will forward this connection to port `#3022` on the destination Node. |
```
Because you pointed the node to the proxy, port 3023 is no longer necessary, because in this case proxy will connect using the outbound tunnel that the node has established. You can safely remove 3023 from the proxy and it will work OK.

inertial-frame replied:
Verified. Missed that when removing ports. Thanks!

klizhentas commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,347 @@
+
+   ```bash
+   # Use tsh to ssh into a Node
+   sudo tsh ssh root@ip-172-31-41-144
````
@inertial-frame you don't need `sudo` for `tsh` to work, so you can safely remove it

klizhentas commented on docs/pages/server-access/getting-started.mdx:
````diff
@@ -0,0 +1,347 @@
+
+   ```bash
+   # Query all Nodes with a label
+   sudo tsh ssh root@env=example
````
`tsh ssh` will work without sudo here (and below too)

klizhentas reviewed (commented):
@inertial-frame we are entering into the final stretch, the structure and diagrams look great, just the right balance of explaining the idea and running through the steps. Last things to address:

* [x] Fix the typo in `doman registrar` to domain registrar
* [x] Remove port 3023 requirement from node
* [x] Remove `sudo` from tsh commands
* [x] Replace ACME-complaint with an alternative, see the comments below

And good to go!

inertial-frame commented:
@klizhentas:

1. Made the requested changes.
2. Made the requested changes for both Guides as well (removed `sudo` from `tsh`).
3. Updated ACME phrasing in article, updated in AWS article too (since it used the previous way), and will update in backport PR's.
4. Verified port number and tsh commands with the changes above.

Thanks for bearing with me through this!

inertial-frame committed:
docs: requested changes

inertial-frame committed:
Merge branch 'branch/v6' into adam/docs/62accsect

inertial-frame committed:
Merge branch 'branch/v6' into adam/docs/62accsect
